File: src/conductances/SKv3_1.py
"""
SKv3_1: Kv3.1-type potassium conductance 

Extracted from BBP NEURON mechanism: SKv3_1.mod
Neuron type: pyramidal neuron

State variables:
    m: activation gate
"""

# euler's constant to 30 digits of precision
e = 2.718281828459045235360287471352

# maximal peak conductance density of channel per unit membrane area (S/cm^2)
gSKv3_1bar = 0.00001

def derivatives(m, V):
    """
    Compute steady-state value and time constant,
    then compute time derivative of the gating variable.

    Inputs:
        - m (float): activation gate variable
        - V (float): membrane potential (mV)

    Returns: 
        - dm_dt (float): time derivative of m
    """

    # steady state activation
    mInf = 1.0 / (1.0 + e**((V - 18.7) / -9.7))

    # activation time constant (ms)
    mTau = 0.2 * 20 / (1+e**((V + 46.560) / -44.140))

    dm_dt = (mInf - m) / mTau

    return dm_dt

def current(m, V, E_K, dt=0.025):
    """
    Compute current based on gating variables and membrane potential.

    Inputs:
        - m (float): activation gate variable
        - V (float): membrane potential (mV)
        - E_K (float): calcium reversal potential (mV)
        - time step (float): time step of derivative measurement

    Returns: 
        - I_K (float): current (mA/cm^2)
    """
    # steady state activation
    mInf = 1.0 / (1.0 + e**((V - 18.7) / -9.7))

    # activation time constant (ms)
    mTau = 0.2 * 20 / (1+e**((V + 46.560) / -44.140))
    
    # cnexp update
    m = mInf + (m - mInf) * e**(-dt/mTau)
    
    gSKv3_1 = gSKv3_1bar * m
    I_K = gSKv3_1 * (V - E_K)
    return I_K

File: src/cpp/SKv3_1.cpp
#include "SKv3_1.hpp"

#include <cmath>

namespace {
const scalar_t e = 2.718281828459045235360287471352;
const scalar_t gSKv3_1bar = 0.00001;
}

scalar_t SKv3_1_derivatives(scalar_t m, scalar_t V) {
    const scalar_t mInf = 1.0 / (1.0 + std::pow(e, ((V - 18.7) / -9.7)));
    const scalar_t mTau = 0.2 * 20.0 / (1.0 + std::pow(e, ((V + 46.560) / -44.140)));

    return (mInf - m) / mTau;
}

scalar_t SKv3_1_current(scalar_t m, scalar_t V, scalar_t E_K, scalar_t dt) {
    const scalar_t mInf = 1.0 / (1.0 + std::pow(e, ((V - 18.7) / -9.7)));
    const scalar_t mTau = 0.2 * 20.0 / (1.0 + std::pow(e, ((V + 46.560) / -44.140)));

    m = mInf + (m - mInf) * std::pow(e, (-dt / mTau));

    const scalar_t gSKv3_1 = gSKv3_1bar * m;
    const scalar_t I_K = gSKv3_1 * (V - E_K);
    return I_K;
}